### ISSUE TITLE
File dependencies from transitive dependencies are stored incorrectly in lock file

### ISSUE THREAD

**Issue by joscha** [yarnpkg/yarn#4388]
**What is the current behavior?**

When having a file dependency that has more file dependencies, these second-level dependencies are referenced in cache, e.g.:
```json
{
  name: "x"
  dependencies: {
    "a": "file:../../a"
  }
}
```
with:
```
{
  name: "a",
  dependencies: {
    "b": "../b"
  }
```
would result in a lock entry like:
```
 "a@file:../../a":
   dependencies:
     "b" "file:../../../../../Library/Caches/Yarn/v1/npm-b"
```
instead of (yarn 0.27.5):
```
 "a@file:../../a":
   dependencies:
     "b" "file:../b"
```

**If the current behavior is a bug, please provide the steps to reproduce.**

Create two chains of dependencies, see: https://github.com/joscha/yarn-101-links especially https://github.com/joscha/yarn-101-links/blob/master/my-package/yarn.lock#L8

**What is the expected behavior?**
Have relative links only, so the whole `node_modules` folder can be gzipped.

**Please mention your node.js, yarn and operating system version.**
node: 8.4.0
yarn 1.0.1
OSX Sierra

**Comment by code0100fun** [comment]
I am seeing this too but it appears that yarn 1.0.1 running on a linux box doesn't have a problem with an OSX `file:../../../../../Library/Caches/Yarn/v1/...` reference. Running install on linux does produce a yarn.lock diff where the caches have been updated to `file:../../../../../.cache/yarn/v1/...`. This doesn't break my CI but it seems like it will cause issues for some people.

**Comment by ZeeCoder** [comment]
We have this issue too.
I don't think it broke anything so far, but yarn.lock changes tend to show up in every MR now.

**Comment by BYK** [comment]
@ZeeCoder which version of Yarn are you using?

**Comment by ZeeCoder** [comment]
1.0.2, had the same with 1.0.1

**Comment by jan-j** [comment]
I'm having same issue with version 1.1.0

**Comment by joscha** [comment]
Still happening on 1.2.1 - this is blocking any upgrade to 1.x - is there any way to escalate this issue?

**Comment by joscha** [comment]
I created a PR with a failing test: https://github.com/yarnpkg/yarn/pull/4717 - still unsure on where the fix has to be made.

**Comment by joscha** [comment]
is there anyone here that would be able to have a look at #4717 and maybe give me a hint on where the fix needs to be?

**Comment by joscha** [comment]
@BYK any way to finish up #4717 and get this resolved?

**Comment by BYK** [comment]
@joscha that was a tough thing to solve due to the internal structure of yarn. I would love to work on it more but my time is limited nowadays so any help appreciated.

**Comment by joscha** [comment]
This issue still exists in 1.5.1

**Comment by billinghamj** [comment]
And in v1.6.0

**Comment by joscha** [comment]
Yes @billinghamj :( I had a stab at finding a fix for this the other week again, but I can't find the right place to make a change for this. This is currently blocking us from updating yarn past 0.27.5

**Comment by joscha** [comment]
Can still see this in 1.7.0

**Comment by johanhenrikssn** [comment]
@joscha Still an issue (in 1.7.0) for us too.

**Comment by AshMcConnell** [comment]
This is an issue for us too with 1.7.0.  Is there a workaround?  Thanks!
